engine/table/src/main/java/io/deephaven/engine/table/impl/multijoin/StaticMultiJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,218 @@
+            this.tableNumber = tableNumber;
+        }
+
+        @Override
+        public void doBuild(RowSequence chunkOk, Chunk<Values>[] sourceKeyChunks) {
+            final long maxSize = numEntries + chunkOk.intSize();
+            slotToOutputRow.ensureCapacity(maxSize);
+            tableRedirSource.ensureCapacity(maxSize);
+            for (WritableColumnSource src : outputKeySources) {
+                src.ensureCapacity(maxSize);
+            }
+            buildFromLeftSide(chunkOk, sourceKeyChunks, tableRedirSource, tableNumber);
+        }
+    }
+
+    protected abstract void buildFromLeftSide(RowSequence rowSequence, Chunk[] sourceKeyChunks,
+            LongArraySource tableRedirSource, long tableNumber);
+
+    protected void buildTable(
+            final BuildContext bc,
+            final RowSequence buildRows,
+            final ColumnSource<?>[] buildSources,
+            final TypedHasherUtil.BuildHandler buildHandler) {
+        try (final RowSequence.Iterator rsIt = buildRows.getRowSequenceIterator()) {
+            // noinspection unchecked
+            final Chunk<Values>[] sourceKeyChunks = new Chunk[buildSources.length];
+
+            while (rsIt.hasMore()) {
+                final RowSequence chunkOk = rsIt.getNextRowSequenceWithLength(bc.chunkSize);
+
+                doRehash(chunkOk.intSize());
+
+                getKeyChunks(buildSources, bc.getContexts, sourceKeyChunks, chunkOk);
+
+                buildHandler.doBuild(chunkOk, sourceKeyChunks);
+
+                bc.resetSharedContexts();
+            }
+        }
+    }
+
+    public void doRehash(final int nextChunkSize) {
+        final int oldSize = tableSize;
+        while (rehashRequired(nextChunkSize)) {
+            tableSize *= 2;
+            if (tableSize < 0 || tableSize > MAX_TABLE_SIZE) {
+                throw new UnsupportedOperationException("Hash table exceeds maximum size!");
+            }
+        }
+        if (tableSize > oldSize) {
+            rehashInternalFull(oldSize);
+        }
+    }
+
+    public boolean rehashRequired(int nextChunkSize) {
+        return (numEntries + nextChunkSize) > (tableSize * maximumLoadFactor);
+    }
+
+    abstract protected void rehashInternalFull(final int oldSize);
+
+    protected int hashToTableLocation(int hash) {
+        return hash & (tableSize - 1);
+    }
+
+    // produce a pretty key for error messages
+    protected String keyString(Chunk[] sourceKeyChunks, int chunkPosition) {
+        return ChunkUtils.extractKeyStringFromChunks(chunkTypes, sourceKeyChunks, chunkPosition);
+    }
+
+    @Override
+    public long getResultSize() {
+        return numEntries;
+    }
+
+    @Override
+    public ColumnSource<?>[] getKeyHashTableSources() {
+        return outputKeySources;
+    }
+
+    @Override
+    public WritableRowRedirection getRowRedirectionForTable(int tableNumber) {
+        return new LongColumnSourceWritableRowRedirection(indexSources.get(tableNumber));
```

Comment:
I think this need not be writable. You just need to mark the redirections in the incremental case to start tracking previous values after the initial build.